### PR TITLE
Add interp

### PR DESCRIPTION
**Pull request by dcherian**
xref #6474

- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -240,6 +241,198 @@ def slices_from_chunks(chunks):
     return list(product(*slices))
 
 
+def subset_to_blocks(
```
Comment (dcherian) [inline dask/array/core.py:244]: Should I move this to slicing.py?

File: dask/array/core.py [hunk 2]
```diff
@@ -240,6 +241,198 @@ def slices_from_chunks(chunks):
...
+    ndim = len(coords)
+    return Array(
+        HighLevelGraph.from_collections(token, graph, dependencies=[data]),
+        token,
+        chunks=(*data._chunks[:-ndim], new_chunks),
+        meta=data,
+    )
+
+
+# Copied from Xarray
```
Comment (dcherian) [inline dask/array/core.py:266]: what's the convention for vendoring functions?

File: dask/array/core.py [hunk 3]
```diff
@@ -240,6 +241,198 @@ def slices_from_chunks(chunks):
...
+    inverse_permutation[indices] = np.arange(len(indices), dtype=np.intp)
+    return inverse_permutation
+
+
+def interp_helper(
+    func: Callable[np.ndarray, ...],
+    data,
+    *,
+    x: tuple[np.ndarray, ...],
+    new_x: tuple[np.ndarray, ...],
```
Comment (dcherian) [inline dask/array/core.py:296]: A nicer signature would be interleaved `x0, new_x0, x1, new_x1`

File: dask/array/routines.py [hunk 4]
```diff
@@ -2558,3 +2558,26 @@ def triu_indices_from(arr, k=0):
...
+
+def np_interp_wrapper(data: np.ndarray, x: np.ndarray, new_x: np.ndarray, **kwargs):
+    for a in (data, x, new_x):
+        assert a.ndim == 1
+    return np.interp(new_x, x, data, **kwargs)
+
+
+@derived_from(np)
+def interp(
+    x: np.ndarray, xp: np.ndarray, fp, left=None, right=None, period=None, **kwargs
```
Comment (dcherian) [inline dask/array/routines.py:2571]: This is a bit limited. It only works if we have `x` and `xp` in memory. but `fp` if potentially a dask array with a number of expensive operations queued up, so perhaps it is still useful.

Comment (phofl) [reply dask/array/routines.py:2571]: We'd have to clearly document this in a doctoring and probably add assertions as well to raise if the first two are dask arrays

File: dask/array/core.py (hunk 2 above)
Comment (phofl) [reply dask/array/core.py:266]: Not really anything, what you have now seems fine

File: dask/array/core.py (hunk 3 above)
Comment (phofl) [reply dask/array/core.py:296]: This is only used internally, correct? If yes, feel free to use the signature that is easier to work with

File: dask/array/core.py [hunk 5]
```diff
@@ -240,6 +241,198 @@ def slices_from_chunks(chunks):
...
+    grouped = tlz.groupby(
+        key=lambda x: tuple(map(int, x[1])), seq=enumerate(needed_chunks)
+    )
+    blocks_to_idx = {
+        block_id: tuple(_[0] for _ in vals) for block_id, vals in grouped.items()
+    }
+    desired_chunks = tuple(len(a) for a in blocks_to_idx.values())
+
+    # subset to needed blocks only
+    subset = subset_to_blocks(
```
[the remark this answers is not in the document]
Comment (phofl) [reply dask/array/core.py:373]: I can help here if needed as well

File: dask/array/tests/test_array_core.py [hunk 6]
```diff
@@ -5406,3 +5406,13 @@ def test_load_store_chunk():
     )
     expected = np.array([])
     assert all(actual == expected)
+
+
+def test_interp():
```
Comment (phofl) [inline dask/array/tests/test_array_core.py:5411]: A few more tests would be great, I'd be happy if you use xarray to call into the functions if that makes things easier fwiw

**Comment by djhoese** [comment]
@dcherian Is there a discussion somewhere that describes why this was closed and work on it was ended? I'm revisiting some old dask/xarray code that is basically doing a `nanquantile` + `interp` and right now wraps `interp` as a Delayed function which is suboptimal dask graph optimization-wise.

**Comment by dcherian** [comment]
It was cleaner to do indexing + dot product : https://github.com/pydata/xarray/issues/6799#issuecomment-2555414780 . Xarray PR welcome!

**Pull request closed without merging**